### ISSUE TITLE
CFBundleDocumentTypes Warning from App Store

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#596]
We got this warning while uploading Moonshine to the App Store today:

>Dear Developer,
>
>We identified one or more issues with a recent delivery for your app, "Moonshine" 2.4.0 (51.0). Your delivery was successful, but you may wish to correct the following issues in your next delivery:
>
>ITMS-90788: Incomplete Document Type Configuration - The CFBundleDocumentTypes dictionary array in the com.moonshine-ide Info.plist should contain an LSHandlerRank value for the CFBundleTypeName '$documentTypeName' entry. Refer to https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/TP40009249-SW1 for more information on the LSHandlerRank key.
>
>After you’ve corrected the issues, you can use Xcode or Application Loader to upload a new binary to App Store Connect.
>
>Best regards,
>
>The App Store Team

This seems to be the relevant section:  https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-101685

**Comment by JoelProminic** [comment]
I think this indicates that "LSHandlerRank" is missing.

>Determines how Launch Services ranks this app among the apps that declare themselves editors or viewers of files of this type. The possible values are: Owner (this app is the primary creator of files of this type), Default (this app is an opener of files of this type; this value is also used if no rank is specified), Alternate (this app is a secondary viewer of files of this type), and None (this app is never selected to open files of this type, but it accepts drops of files of this type). Launch Services uses the value of LSHandlerRank to determine the app to use to open files of this type. The order of precedence is: Owner, Default, Alternate. This key is available in macOS 10.5 and later and iOS 3.0 and later.

I confirmed we do not have this in info.plist currently:

        <key>CFBundleDocumentTypes</key>
        <array>
            <dict>
                <key>CFBundleTypeExtensions</key>
                <array>
                    <string>as</string>
                </array>
                <key>CFBundleTypeMIMETypes</key>
                <array>
                    <string>text/plain</string>
                </array>
                <key>CFBundleTypeName</key>
                <string>Moonshine ActionScript File</string>
                <key>CFBundleTypeRole</key>
                <string>Editor</string>
            </dict>

If we need to add LSHandlerRank, I suggest what we set these types to "Owner":

    veditorproj
    javaproj
    grailsproj


And these to "Alternate" since there are other applications that use them.  I think the only one we could reasonably set to "Default" would be asproj.

    as
    css
    txt
    mxml
    xml
    js
    htm
    html
    awd


We'll need to retest opening these files from the Finder after this update.

**Comment by rat-moonshine** [comment]
Build script updated to generate `LSHandlerRank` tag against file-type-associations accordingly:

Type/Extension | LSHandlerRank-value
--------------------- | ------------
as3proj | Default
veditorproj | Owner
javaproj | Owner
grailsproj | Owner
as, css, txt, mxml, xml, js, htm, html, awd | Alternate

Tested on macOS, double-clicking on any of the supported project-configuration file opens the project to Moonshine. Files those marked as 'Alternate' generally opens by other default-programs as set by the OS when double-clicked, but an 'Open with Moonshine' option available to those files.

**Comment by JoelProminic** [comment]
I confirmed that these settings are in place for both the non-Sandbox and AppStore versions.

I think this issue was left open so that we could test the behavior for opening the different file types.  I'll see if I can do some quick tests.

**Comment by JoelProminic** [comment]
I tested with the non-Sandbox build to avoid sandbox errors.

Type | Behavior
-----|---------
as3proj | Opens in Moonshien by default
veditorproj | Not the default but appears in the list (I think I set another application as the default previously)
javaproj | Opens in Moonshine by default
grailsproj | Opens in Moonshine by default
as, css, txt, mxml, xml, js, htm, html, awd (test some examples) | Shows up in Open With list, and enabled in "Other" filechooser.

This looks fine to me.
